Add a basic keyboard shortcuts overview dialog

kituuu opened the pull request:
#11711 [REVIEWs & CHANGEs]

The basic structure for this issue is ready. 

TODO (For my convenience):
1. Write unit tests
2. Get maintainers' feedback for changes or improvements.
[
Check this](https://github.com/wagtail/wagtail/issues/11711#issuecomment-1974819266)

Do you think the implementation of the design from Figma is necessary? Because right now, the default styling and existing dialogue template do not fit that design. I will have to write separate CSS and a new dialogue box (Maybe we can make this new one default if it is better visually).
[Personally I feel this dialogue box design (figma one) is a bit better in distinguishing different sections : D ]

lb- commented on wagtail/admin/templates/wagtailadmin/shared/keyboard_shortcuts_dialog.html:
```diff
@@ -0,0 +1,58 @@
+    <table class="w-h-full w-overflow-y-scroll">
+        <caption class="w-sr-only">
+            All keyboard shortcuts
+        </caption>
```
Just a reminder to use translated values for all text.

kituuu replied:
Ah, sry, but what do you mean by translated values?

lb- replied:
No problems. Essentially we want to make sure that all strings / text is marked for translation. 

https://docs.wagtail.org/en/stable/contributing/translations.html#marking-strings-for-translation

This leverages a Django system of translation. 

I don't think we need to mark the actual shortcuts for translation but I'll check with the translation team.

lb- replied:
No need to pass the shortcut keys label (e.g. `Ctrl + z`) through the translation system on this PR. 

In the future we may review how these labels are generated once we start to consider more complex customisation use cases.

lb- commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1296,6 +1297,15 @@ def workflow_status_with_date(workflow_state):
+    user_agent = context["request"].headers.get("User-Agent", "")
+    is_mac = re.search(r"Mac|iPod|iPhone|iPad", user_agent)
+    return {
+        "meta": "⌘" if is_mac else "Ctrl",
```
```suggestion
        "mod": "⌘" if is_mac else "Ctrl",
```

Just realised we should be calling this `mod` not `meta` - mod seems to be the convention for things like this. 

The template would also need to be updated.

kituuu commented:
<img width="951" alt="image" src="https://github.com/wagtail/wagtail/assets/66073214/ff3a6263-1f20-41c3-9fd5-30a3d60c0dc2">


Hey @lb- , I have tried my best with the design. I used a custom scss file for styling. And also, instead of hard coding everything, I just made a dictionary of all the shortcuts so that in the future, addition and removal of shortcuts could be more accessible. I am not good in naming CSS classes LoL : D

Also, the figma design has the tables divided into 2 columns but as we have less width available in our current dialog, ig it's better to keep single column.

If everything is fine, please tell me more about writing unit tests for this PR.

lb- commented:
Awesome I will take a look.

Just a reminder to review the failing CI items when you get the chance, some may be just formatting (see formatting guidelines https://docs.wagtail.org/en/stable/contributing/python_guidelines.html#python-coding-guidelines - `make format` should be all that's needed though).

As for unit tests, it looks like we have lots of tests that are globally checking for `td`/`tr` in responses, and the addition of this table is breaking those tests. I might have a look at these and see if there is a simple solution to suggest.

```

======================================================================
FAIL: test_simple (wagtail.admin.tests.viewsets.test_model_viewset.TestUsageView)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/wagtail/wagtail/wagtail/admin/tests/viewsets/test_model_viewset.py", line 1088, in test_simple
    self.assertEqual(len(tds), 3)
AssertionError: 19 != 3

======================================================================
FAIL: test_usage_with_describe_on_delete (wagtail.admin.tests.viewsets.test_model_viewset.TestUsageView)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/wagtail/wagtail/wagtail/admin/tests/viewsets/test_model_viewset.py", line 1158, in test_usage_with_describe_on_delete
    self.assertEqual(len(tds), 3)
AssertionError: 19 != 3

======================================================================
FAIL: test_usage_without_permission_on_referrer (wagtail.admin.tests.viewsets.test_model_viewset.TestUsageView)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/wagtail/wagtail/wagtail/admin/tests/viewsets/test_model_viewset.py", line 1136, in test_usage_without_permission_on_referrer
    self.assertEqual(len(tds), 3)
AssertionError: 19 != 3

======================================================================
FAIL: test_disable_preview_on_revisions_list (wagtail.snippets.tests.test_preview.TestDisablePreviewWithEmptyModes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/wagtail/wagtail/wagtail/snippets/tests/test_preview.py", line 474, in test_disable_preview_on_revisions_list
    self.assertNotContains(response, "Preview")
  File "/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/django/test/testcases.py", line 676, in assertNotContains
    self.assertEqual(
AssertionError: 1 != 0 : Response should not contain 'Preview'

======================================================================
FAIL: test_disable_preview_on_revisions_list (wagtail.snippets.tests.test_preview.TestDisablePreviewWithoutMixin)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/wagtail/wagtail/wagtail/snippets/tests/test_preview.py", line 474, in test_disable_preview_on_revisions_list
    self.assertNotContains(response, "Preview")
  File "/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/django/test/testcases.py", line 676, in assertNotContains
    self.assertEqual(
AssertionError: 1 != 0 : Response should not contain 'Preview'

======================================================================
FAIL: test_custom_columns (wagtail.snippets.tests.test_viewset.TestListViewWithCustomColumns)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/wagtail/wagtail/wagtail/snippets/tests/test_viewset.py", line 721, in test_custom_columns
    self.assertTagInHTML("<th>", html, count=7, allow_extra_attrs=True)
  File "/home/runner/work/wagtail/wagtail/wagtail/test/utils/wagtail_tests.py", line 208, in assertTagInHTML
    self.assertEqual(
AssertionError: 30 != 7 : Found 30 instances of '<th>' in response (expected 7)

----------------------------------------------------------------------
```

lb- commented:
@kituuu for the tests that have the missing counts, you can rebase / merge in the changes from this PR to your PR https://github.com/wagtail/wagtail/pull/11724

For the preview link tests, here's a PR that will hopefully make those tests not fail with your PR including the word 'preview' in the dialog https://github.com/wagtail/wagtail/pull/11725

kituuu commented:
Okay, I will try to fix all failed tests and stuff after my college. Meanwhile if u r free can u also drop some review on this PR #11714.

kituuu commented:
I guess some test fails are still there even after running make format and merging those 2 PRs.

laymonage commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1301,6 +1302,55 @@ def workflow_status_with_date(workflow_state):
+    is_mac = re.search(r"Mac|iPod|iPhone|iPad", user_agent)
+    shortcuts_category = {
+        "Actions": [
+            {"Label": "Save", "Shortcut": ("⌘" if is_mac else "Ctrl") + " + s"},
```
Instead of repeating the logic, we can extract this to a variable, e.g.

```py
modifier = "⌘" if is_mac else "Ctrl"
```

and in this case I think the keys should be in snake case, e.g.

```suggestion
            {"label": "Save", "shortcut": f"{modifier} + s"},
```

kituuu replied:
Done

kituuu commented:
@lb- Can you suggest something for this PR too? So that I can try to modify some tests to support this PR. : D

lb- reviewed (changes requested):
@kituuu  this is awesome progress so far, thank you!

I have added a bunch of comments inline, there will probably be a bit to do after that (we may want to tweak styles, review tests to add) but overall this is on the right track.

Items to action from this feedback;

1. [x] Update classes to use the `w-prefixed-kebab-convention` and avoid using `&___some-selector` and instead re-write the full name in each block.
2. [x] Use `theme` variables for any sizing, font weight, widths/padding/margins etc, if a variable does not exist, use the closest to your design goal. This ensures consistency through the UI and that we are using dark/light theme variations correctly + rems consistently.
3. [x] Translations; every single string needs to be translated, except the actual shortcuts, this is in the template HTML and the template tags.
4. [x] Remove the duplicate column header for the table
5. [x] For the template tag, use a tuple as a key with an identifier (simple kebab-string) and the translated label.
6. [ ] Use the same variable name as the value passed in the context from the template tag for code readability
7. [ ] Move the 'common actions' above 'actions' & rename 'actions' to 'Page editing actions'
8. [ ] Move the position, in the code, of the new `register_keyboard_shortcuts_menu_item` to a more intuitive position in the code

One column for this layout is fine I think we do not want to have to redesign the whole dialogs as part of this initial work.

I have not looked at the tests sorry but encourage you to have a full read of the CI output to try to dig into what the cause could be.

If you can, I would suggest a rebase/squash on main after these updates are done, it will help the next round of reviews.

Thank you again! this is great to see.

@thibaudcolas  - please take a look if you get the chance also. Thanks @laymonage  for your input.

kituuu commented:
> @kituuu this is awesome progress so far, thank you!
> 
> I have added a bunch of comments inline, there will probably be a bit to do after that (we may want to tweak styles, review tests to add) but overall this is on the right track.
> 
> Items to action from this feedback;
> 
>     1. [ ]  Update classes to use the `w-prefixed-kebab-convention` and avoid using `&___some-selector` and instead re-write the full name in each block.
> 
>     2. [ ]  Use `theme` variables for any sizing, font weight, widths/padding/margins etc, if a variable does not exist, use the closest to your design goal. This ensures consistency through the UI and that we are using dark/light theme variations correctly + rems consistently.
> 
>     3. [ ]  Translations; every single string needs to be translated, except the actual shortcuts, this is in the template HTML and the template tags.
> 
>     4. [ ]  Remove the duplicate column header for the table
> 
>     5. [ ]  For the template tag, use a tuple as a key with an identifier (simple kebab-string) and the translated label.
> 
>     6. [ ]  Use the same variable name as the value passed in the context from the template tag for code readability
> 
>     7. [ ]  Move the 'common actions' above 'actions' & rename 'actions' to 'Page editing actions'
> 
>     8. [ ]  Move the position, in the code, of the new `register_keyboard_shortcuts_menu_item` to a more intuitive position in the code
> 
> 
> One column for this layout is fine I think we do not want to have to redesign the whole dialogs as part of this initial work.
> 
> I have not looked at the tests sorry but encourage you to have a full read of the CI output to try to dig into what the cause could be.
> 
> If you can, I would suggest a rebase/squash on main after these updates are done, it will help the next round of reviews.
> 
> Thank you again! this is great to see.
> 
> @thibaudcolas - please take a look if you get the chance also. Thanks @laymonage for your input.

@lb- It's my pleasure to work with you guys. I will make all the changes as soon as possible. 🙃

kituuu commented on client/scss/components/_keyboard-shortcuts.scss:
```diff
@@ -0,0 +1,32 @@
+.keyboard_shortcut {
```
Okay, I will keep that in mind for future issues. I read the guidelines and tried using BEM architecture (&__somethings). Maybe I missed the w- part.

lb- replied:
You were using the right classes for BEM, thank you. It's a bit confusing I know as we only started using w at the start in the last year or so. Some code doesn't yet use this. 

Additionally, the BEM architecture is fine to use without the nested scss creation of classnames. The idea of just writing the whole name comes from making the code easier to search through. It's a much newer guideline, last 6 months. 

Thanks for understanding, the CSS in Wagtail is constantly evolving and still has lots of older approaches in the code.

kituuu commented on client/scss/components/_keyboard-shortcuts.scss:
```diff
@@ -0,0 +1,32 @@
+    font-weight: 400;
+    margin-bottom: 0.75rem;
+    margin-top: 1.2rem;
+  }
```
Done; as I have removed nesting, I gave two new class names to label and key th & td because id nesting is removed. This might affect other th td in the DOM.

lb- replied:
It should be fine, as the classname is unique to the ones in this new template only.

kituuu commented:
Hey, @lb-, I have made all the changes you asked for. I will now try to figure out the failed test cases.

lb- commented:
@kituuu I know that debugging unit tests can be a bit tricky but I will walk you through one.

*`  File "/home/runner/work/wagtail/wagtail/wagtail/admin/tests/viewsets/test_model_viewset.py", line 1052, in test_empty
    self.assertIsNone(table)`
* The error is saying 'table....' is not none.
* The unit test is in the file `test_model_viewset.py` on line 1052 
https://github.com/wagtail/wagtail/blob/75af2f22beab7f905c9b8212f7092c01a9fe1cc9/wagtail/admin/tests/viewsets/test_model_viewset.py#L1043-L1052
* In that test we can see `table = soup.select_one("table")` is trying to find any element with type `table`, however, due to your change there will now be a new table, the one for the keyboard shortcuts.
* This `soup.select_one` is the Soup library https://beautiful-soup-4.readthedocs.io/en/latest/ and `select_one` accepts a query selector.
* So what we need to do is ensure this test only looks for the correct `table` and finds that it does not exist, in our case we could update this line to `table = soup.select_one("#listing-results table")`.
* This means it will look through the HTML and only find the `table` within the element with ID `listing-results.
* Once you get this test passing, you will find that `wagtail.admin.tests.viewsets.test_model_viewset.TestHistoryView.test_filtered_no_results`, `wagtail.admin.tests.viewsets.test_model_viewset.TestUsageView.test_empty` will be very similar.

```py
    def test_empty(self):
        ModelLogEntry.objects.all().delete()
        response = self.client.get(self.url)
        soup = self.get_soup(response.content)
        results = soup.select_one("#listing-results")
        table = soup.select_one("#listing-results table")
        self.assertEqual(response.status_code, 200)
        self.assertIsNotNone(results)
        self.assertEqual(results.text.strip(), "There are no log entries to display.")
        self.assertIsNone(table)
```


Further tests
* For `wagtail.admin.tests.viewsets.test_model_viewset.TestHistoryView.test_filters` -> this is very similar, notice there is a count of rows using a selector for ALL rows in any table. `rows = soup.select("tbody tr")`.
* `wagtail.snippets.tests.test_preview.TestDisablePreviewWithEmptyModes.test_disable_preview_on_revisions_list` & `wagtail.snippets.tests.test_preview.TestDisablePreviewWithoutMixin.test_disable_preview_on_revisions_list` -> These look like the same problem where the word Preview is being searched for in the HTML response content but the new keyboard shortcut content is returning the word Preview. Maybe remove the check `self.assertNotContains(response, "Preview")` or refine that text label to be more specific.

kituuu commented:
Hey @lb-, only one following wedding is left. Can you say something regarding this.
<img width="1721" alt="image" src="https://github.com/wagtail/wagtail/assets/66073214/b00e03af-ce21-4c54-ac0f-d94b80c72ef7">

lb- commented on wagtail/admin/templates/wagtailadmin/shared/keyboard_shortcuts_dialog.html:
```diff
@@ -0,0 +1,30 @@
+{% load wagtailadmin_tags i18n %}
+{% dialog icon_name='regex' id='keyboard-shortcuts-dialog' title=_("Keyboard shortcuts") %}
+    {% if shortcuts%}
```
I would not worry about this logic here, as we are currently hard-coding the shortcuts from the template tag there should never be a case where there are no shortcuts.

lb- commented on wagtail/admin/templates/wagtailadmin/shared/keyboard_shortcuts_dialog.html:
```diff
@@ -0,0 +1,30 @@
+                        <th class="w-keyboard-shortcuts__category" scope="rowgroup">{{ category.1 }}</th>
+                    </tr>
+                    {% for shortcut in shortcuts %}
+                        <tr class="w-keyboard_shortcuts__key ">
```
```suggestion
                        <tr class="w-keyboard_shortcuts__key">
```

Just a small class whitespace cleanup

lb- commented:
Here's a starting point for Python unit tests. @kituuu 

File: `wagtail/admin/tests/tests.py` - add the new test block at the end.

```python

class TestKeyboardShortcutsDialog(WagtailTestUtils, TestCase):
    def setUp(self):
        self.login()

    def test_keyboard_shortcuts_trigger_in_sidebar(self):
        response = self.client.get(reverse("wagtailadmin_home"))
        self.assertEqual(response.status_code, 200)

        sidebar_data = (
            self.get_soup(response.content)
            .select_one("#wagtail-sidebar-props")
            .contents[0]
        )

        self.assertIn(
            json.dumps(
                {
                    "role": "button",
                    "data-a11y-dialog-show": "keyboard-shortcuts-dialog",
                    "data-action": "w-action#noop:prevent:stop",
                    "data-controller": "w-action",
                }
            ),
            sidebar_data,
        )

    def test_keyboard_shortcuts_dialog(self):
        response = self.client.get(reverse("wagtailadmin_home"))

        self.assertEqual(response.status_code, 200)
        self.assertTemplateUsed(
            response, "wagtailadmin/shared/keyboard_shortcuts_dialog.html"
        )

        soup = self.get_soup(response.content)

        # Check that the keyboard shortcuts dialog is present
        shortcuts_dialog = soup.select_one("#keyboard-shortcuts-dialog")
        self.assertIsNotNone(shortcuts_dialog)

        # Check that the keyboard shortcuts dialog has basic accessible content
        self.assertIn(
            "All keyboard shortcuts", shortcuts_dialog.find("caption").prettify()
        )
        self.assertIn("Keyboard shortcut", shortcuts_dialog.find("thead").prettify())

```

Note that we have to use `prettify` to get the text content, this is due to the text being within a `template` element which BeautifulSoup will correctly ignore.

This unit test set should be updated to also check a few other parts of the shortcuts, and you will also need to test that the cmd / control key is correctly identified based on user agent. The default test will use `Ctrl`, so you will need to add a separate test that emulates MacOS for the different keyboard `cmd` symbol.

See https://docs.djangoproject.com/en/5.0/topics/testing/tools/#making-requests for how to go about this. Something like...

```python
# In the test where we want to test the non-default behaviour we will not use `self.client` instead create a new client that has a different user agent.
client = Client(headers={"user-agent": "something-that-will-be-like-macos"})
client.get(reverse("wagtailadmin_home")
```

Good luck, hopefully that helps you with the next step.

kituuu commented:
Hey, @lb- So, firstly, I added the two tests you gave me. Apart from this, I am writing some tests to simulate the behaviour of Mac & windows machines to test the shortcuts. 

My initial plan was to create a new Client object with some header content resembling the Marcos 
`client = Client(
            headers={
                "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36"
            }
        )`
Then, get the table and kbd and check whether the correct keys are present in kbds.
But with this, I faced one issue.
When I ran `response = client, get (reverse("wagtailadmin_home"))`, I did not get the desired result (which we got in the above two tests). Instead, I got some response with `"URL":/admin/login/?next=/admin/` through which I understand that in our class in setup, we ran self.login() I guess which login the test user from `WagtailTestUtils` but unfortunately our new Client is out of the scope of WagtailTestUtils, so every time I use the client. Get, I got a 302 status code.

Things I tried to fix this:
1. I used the client. Login and client.forced_login methods (I thought it would do something, but it doesn't).
2. I tried assigning 
```
self.client =  Client(
            headers={
                "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36"
            }
        )
Self. login
```
But something else is needed.
3. Then I changed my plan. I thought of checking directly the user_agent and `is_mac = re.search(r"Mac|iPod|iPhone|iPad", user_agent)` because ultimately our template is decided based on user_agent & is_mac variable... I tried extracting the user_agent from the request of `response = client, get (reverse("wagtailadmin_home"))`, but it doesn't have the user-agent header... I tried logging in using a post request once, and the request post request had this user-agent header. So, I used that in my unit tests.

<img width="1728" alt="Screenshot 2024-03-14 at 12 47 16 AM" src="https://github.com/wagtail/wagtail/assets/66073214/ead787ed-5567-4dd0-8e85-979a73d5e503">


If you have a better way of doing this, do tell me. If I have to write some more unit tests, then also tell me.

lb- commented:
Ahh I see, yeah, I did not thank of that, great digging. @kituuu 

One option would be to just change the `self.client` to your client within the test and login.

```
    def test_mac_behavior(self):
        user = self.create_user()
        self.client = Client(...)
        self.login()

        response = #...
```


As an aside, it's much easier to work with pasted text from your code in comments and not a screenshot, saves me having to re-write what you have done so far.

kituuu commented:
@lb- So I tried the way you suggested.

```
    def test_mac_behavior(self):
        macuser = self.create_user(username='macuser',email="macuser@test.com", password='macuser')
        self.client = Client(
            headers={
                "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36"
            }
        )
        self.login(user=macuser, username='macuser', password='macuser')
        response = self.client.get(reverse("wagtailadmin_home"))
        print(response)

```

On running this, I got 
`<HttpResponseRedirect status_code=302, "text/html; charset=utf-8", url="/admin/login/?next=/admin/">` 
(i.e. the response, which means this is also not working)...

But with this, I just got the idea of trying it in the setUp function itself. So I created a new class. 

```
class TestMacKeyboardShortcutsDialog(WagtailTestUtils, TestCase):
    def setUp(self):
        # Creates a client with a Mac user agent
        self.client = Client(
            headers={
                "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36"
            }
        )
        self.login()

    def test_mac_useragent_and_behavior(self):
        response = self.client.get(reverse("wagtailadmin_home"))

        # Check that the user agent is a Mac
        user_agent = response.context["request"].headers.get("User-Agent", "")
        is_mac = re.search(r"Mac|iPod|iPhone|iPad", user_agent)

        # Add assertions based on expected Mac behavior
        self.assertTrue(is_mac)

        # Check that the keyboard shortcuts dialog has Mac-specific content
        soup = self.get_soup(response.content)
        shortcuts_dialog = soup.select_one("#keyboard-shortcuts-dialog")
        all_shortcuts = shortcuts_dialog.select("kbd")
        for shortcut in all_shortcuts:
            # All shortcuts should have the ⌘ symbol
            self.assertIn("⌘", shortcut.prettify())

```
And this second one is working as expected.
I don't know why it isn't working in the first case. 
I have pushed the code.

lb- commented:
@kituuu  great, that's fine to do the second class approach.

However, may I suggest you move BOTH test classes to their own file instead.

So instead of putting the tests in `wagtail/admin/tests/tests.py`, please create a new file `wagtail/admin/tests/test_keyboard_shortcuts.py` and put both there.

It's likely we will add to this keyboard shortcuts system over time, including the ability to support a hooks or some other mechanism for registering shortcuts so having its own file makes sense. This feature will be a similar scale to dismissables which also has its own test file `wagtail/admin/tests/test_dismissibles.py`

Great investigating here, I have learned a lot from you on this also.

kituuu commented:
Done

lb- commented:
Thanks. I'll do a full review and see if there is anything missing. 

Don't worry about quantity of contributions, this is awesome work and will be a great improvement for all Wagtail users.

allcaps commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1296,6 +1297,56 @@ def workflow_status_with_date(workflow_state):
+                "shortcut": f"{modifier} + shift + v",
+            },
+            {"label": _("Undo"), "shortcut": f"{modifier} + z"},
+            {"label": _("Redo"), "shortcut": f"{modifier} + y"},
```
The default Redo on Mac is `⌘ + SHIFT + Z`, this one opens up history in Chrome. ;)

allcaps replied:
If we use words for modifier keys like SHIFT, we need to check if these words need translation.
So maybe it is better to stick with icons?

> EN: Press the shift key.
> FR: Appuyez sur la touche Maj.

lb- replied:
I already checked that with the translations channel, they advised against this.

allcaps commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1296,6 +1297,56 @@ def workflow_status_with_date(workflow_state):
+            {"label": _("Cut"), "shortcut": f"{modifier} + x"},
+            {"label": _("Paste"), "shortcut": f"{modifier} + v"},
+            {
+                "label": _("Paste without formatting"),
```
The mac wording for this label is "Paste and Match Style".

Not sure how close we want to stay to OS style icons/wording.

lb- replied:
How about `Paste text only`?

allcaps replied:
I think Windows-speak is "Paste without formatting" and on Mac it is "Paste and Match Style".
The feature seems the same: Whatever snippet is in the clipboard, drop the styling, but match the current style in the text editor.

I assume (don't have Windows here), that if you use "Paste without formatting" into a heading element, the text still gets the heading elements style applied???

My suggestion is, to align shortcuts and labels (wording) with the OS.

Something like:

``` python

mac = "mac"
win = "windows"
os = mac if re.search(r"Mac|iPod|iPhone|iPad", user_agent) else win

shortcuts = {
    ("common-actions", _("Common Actions")): [
        ...,
        {
            "label": _("Paste and Match Style") if os == mac else _("Paste without formatting"),
            "shortcut": f"{modifier} + shift + v",
        },
        # You'd need something similar for the other shortcut.
        {
            "label": _("Redo"),
            "shortcut": f"Shift + {modifier} + z"  if os == mac else f"{modifier} + y",
        },
        ...,
    ]
}
```

allcaps replied:
This ^ can become quite ugly code. So we can also do something like this:
``` python
# Windows is the default.
shortcuts = [
    # Win/Mac are the same.
    {
        "label": "Some label",
        "shortcut":  "Some shortcut",
    },
    # Mac has other label
    {
        "label": "Some label",
        "label_mac": "Some Mac label",
        "shortcut":  "Some shortcut",
    },
    # Mac has other shortcut
    {
        "label": "Some label",
        "shortcut": "Some shortcut",
        "shortcut_mac": "Some Mac shortcut",
    },
    # Mac has other label and shortcut
    {
        "label": "Some label",
        "label_mac": "Some Mac label",        
        "shortcut": "Some shortcut",
        "shortcut_mac": "Some Mac shortcut",
    },
]

# Update the labels and shortcuts with the Mac variant if we are on OSX or iOS.
if re.search(r"Mac|iPod|iPhone|iPad", user_agent):
    for item in shortcuts:
        item["label"] = item.pop("label_mac", item["label"])
        item["shortcut"] = item.pop("shortcut_mac", item["shortcut"])


print(shortcuts)
```

kituuu replied:
@lb- @allcaps I understand the required changes, but do we have any list or thing for these changes? For example, what should the label name be on Mac and Windows? Right now, I only know that "Paste and Match Style" and "Paste without formatting" need to be changed, but I don't know about other shortcuts.

kituuu replied:
I have made changes to this specific "Paste and Match Style". Are there any more labels that need to be changed based on OS?

kituuu replied:
Also, there should be some changes in unit tests to check this change. Should I update the unit tests in this PR itself? Or can we ignore it for now?

kituuu replied:
@lb- : D

allcaps commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1296,6 +1297,56 @@ def workflow_status_with_date(workflow_state):
+            {"label": _("Redo"), "shortcut": f"{modifier} + y"},
+        ],
+        ("page-editing-actions", _("Page editing actions")): [
+            {"label": _("Save"), "shortcut": f"{modifier} + s"},
```
I didn't know about `⌘ + s` :clap:

Models (snippets) are saved, but pages `Save draft`. Maybe the label should be `Save (draft)`?

This shortcut overrides the default browser behaviour. And in this specific case, that seems okay. Saving HTML is not a common thing to do. Overriding OS/browser shortcuts is something we should consider when adding more shortcuts.

This shortcut doesn't work when the cursor is inside a text-area. It will use the default, trigger the browser to save the HTML. -- Fixing this might be outside the scope of this PR.

lb- replied:
Great feedback - thanks!

The behaviour for keyboard shortcuts to not listen inside input fields is intentional. However, I see how this could be confusing for the CMD + S scenario. 

I'll look at that for a different PR.

Generally I agree that we should not override keyboard shortcuts in the browser but this has been in the code for a long time. Be sure to have a read of https://github.com/wagtail/wagtail/issues/3949 for the big picture plan for keyboard shortcuts, I've already raised the same concerns there.

allcaps commented on wagtail/admin/wagtail_hooks.py:
```diff
@@ -998,6 +998,23 @@ def register_editors_guide_menu_item():
+        },
+        name="keyboard-shortcuts-trigger",
+        url="#",
+    )
```
It would be very cool if opening the dialogue could be a shortcut too. `command + /` is common.

lb- replied:
Yep. It's on the plans in https://github.com/wagtail/wagtail/issues/3949 - it's dependant on a different bit of work making it easier to add keyboard shortcuts with just data attributes. 

See https://github.com/wagtail/wagtail/issues/10167 & https://github.com/wagtail/wagtail/pull/11740

lb- commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1296,6 +1297,56 @@ def workflow_status_with_date(workflow_state):
+            {"label": _("Redo"), "shortcut": f"{modifier} + y"},
+        ],
+        ("page-editing-actions", _("Page editing actions")): [
+            {"label": _("Save"), "shortcut": f"{modifier} + s"},
```
```suggestion
            {"label": _("Save (draft)"), "shortcut": f"{modifier} + s"},
```

allcaps replied:
This is really a nitpick. For the first iteration we can go with "Save".

Just to explain myself: "Save changes"  or "Save" do not clarify that pages don't have a "Save" option. Pages have a workflow with "Save draft" and "Publish". Therefore, "Save" on a page will always be result in a "Save draft".  I think some users expect saving a already publish page will publish their changes. But that won't happen. They just get a new draft.

The modal with shortcuts is like a cheatsheet. And it is common for cheatsheets to have additional explanation. Maybe help texts? "A pages will save a draft." 

Nice to have. Not for now.

lb- replied:
Yep. Let's keep this simple for now, I anticipate we would have a way for the label/shortcut to be registered somehow at a View or Form level. 

Or we would drive this from the JS side and pull out the label from the element itself. 

As this is static for this first version I think we'll have to compromise with something generic. 

But if you have ideas on how we could provide a way for this to be registered - I'm super keen to hear them. Add a comment to https://github.com/wagtail/wagtail/issues/3949

kituuu replied:
Done

lb- commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1296,6 +1297,56 @@ def workflow_status_with_date(workflow_state):
+                "shortcut": f"{modifier} + shift + v",
+            },
+            {"label": _("Undo"), "shortcut": f"{modifier} + z"},
+            {"label": _("Redo"), "shortcut": f"{modifier} + y"},
```
```suggestion
            {"label": _("Redo"), "shortcut":  f"{modifier} + Shift + z" if is_mac else f"{modifier} + y"},
```

lb- commented on wagtail/admin/templates/wagtailadmin/shared/keyboard_shortcuts_dialog.html:
```diff
@@ -0,0 +1,28 @@
+{% load wagtailadmin_tags i18n %}
+{% dialog icon_name='regex' id='keyboard-shortcuts-dialog' title=_("Keyboard shortcuts") %}
```
I suggest we put the root `classname="w-keyboard-shortcuts"`  on the dialog here, instead of on the `table`. Will make it easier to style the content as a whole.

lb- commented on wagtail/admin/templates/wagtailadmin/shared/keyboard_shortcuts_dialog.html:
```diff
@@ -0,0 +1,28 @@
+{% load wagtailadmin_tags i18n %}
+{% dialog icon_name='regex' id='keyboard-shortcuts-dialog' title=_("Keyboard shortcuts") %}
+    <table class="w-keyboard-shortcuts">
```
```suggestion
    <table class="w-w-full">
```

Probably simpler to have the full width as a tailwind class, with the root class moved to the dialog.

lb- commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -1296,6 +1297,61 @@ def workflow_status_with_date(workflow_state):
+    user_agent = context["request"].headers.get("User-Agent", "")
+    is_mac = re.search(r"Mac|iPod|iPhone|iPad", user_agent)
+    modifier = "⌘" if is_mac else "Ctrl"
+    shortcuts = {
```
Sorry for the back and forth on structure here but a few changes I will do when merging but wanted to explain.

1. Let's use tuples for the shortcuts ->  `{"label": _("Copy"), "shortcut": f"{modifier} + c"}` would be `(_("Copy"), f"{modifier} + c")`. This is a bit easier to read when looking at a lot of these.
2. Let's stick to the same casing for `Shift` not a mix of `Shift` and `shift`.
3. Let's stick to the same casing for titles and labels, e.g. `Common actions` not `Common Actions` and `Paste and match style` not `Paste and Match Style`. Leaning towards 'Sentence case' not 'Title Case'.
4. Might be nice to add a simple docstring on this class for explanation.

```
    """
    Renders the keyboard shortcuts dialog content with the
    appropriate shortcuts for the user's platform.
    Note: Shortcut keys are intentionally not translated.
    """
```

lb- commented on wagtail/admin/wagtail_hooks.py:
```diff
@@ -998,6 +998,23 @@ def register_editors_guide_menu_item():
 
 
+@hooks.register("register_help_menu_item")
+def register_keyboard_shortcuts_menu_item():
```
Let's add a simple docstring here, helps for future context. I can add this when merging.

```suggestion
def register_keyboard_shortcuts_menu_item():
    """
    Triggers the keyboard shortcuts dialog to open when clicked
    while preventing the default link click action.
    """

```

lb- reviewed (approved):
Thank you @kituuu  and also thanks for the extra cross-check @allcaps .

I have reviewed the shortcuts and designs and will make some changes when merging to get this finalised.

For the styling, we just needed some extra refinement to get this closer to the design. Font sizes/weights, colours of headings, background on the `kbd`. I have also simplified the custom classes accordingly. Your work here is great but just wanted to get that last bit of polish in.

I have a few other recommendations but these are quite small and nitpicky so I will just apply these when merging. Tuples vs dicts, adding Docstrings and comments. I have renamed 'page editing actions' to just 'Actions' again, as this can be used for Snippets and potentially other places also.

Finally, I have opted to include a `{% block shortcuts %}` within the template. We do not want to officially document this but at least there is an indirect way to customise these shortcuts (simply) when this goes live.

Here's the final state (comparing to the designs).

It's not exactly what you envisioned @benenright  but as close as we can get without redesigning parts of the dialog structure and without introducing more complex HTML for the keyboard values (e.g. you have bold for the letters in the shortcuts).

![Screenshot 2024-03-16 at 4 38 47 pm](https://github.com/wagtail/wagtail/assets/1396140/e3a388ac-1970-4813-9952-3013d07dad53)


I have tested in Chrome 124, Safari 17.1 and Firefox 123 - looking good in all.

I will make the changes discussed, force push with a changelog and look to merge this in.

lb- commented:
https://github.com/wagtail/wagtail/issues/11767 created for the next bit to do (add an icon), for anyone who is keen.

kituuu commented:
Thanks @lb-. Do you have some other issues for me to work on?
